### ISSUE TITLE
.dm3/.dm4 reader should be advertised on website

### ISSUE THREAD

**Issue by thomasaarholt** [hyperspy/hyperspy#1677]
I've just discovered the existing dm-plugin for reading and writing hdf5: https://github.com/niermann/gms_plugin_hdf5/releases

It has previously been discussed here: #793 

It works quite well, but I believe it is only compiled for 64-bit DM at the moment. It would be nice if it could be advertised on the hyperspy website.

**Comment by magnunor** [comment]
It should be compiled for 32-bit as well, judging by the files present in the release zip: 
- GMS-1.X (32 bit): hdf5_GMS1X_x86.dll
- GMS-2.X (32 bit): hdf5_GMS2X_x86.dll
- GMS-2.X (64 bit): hdf5_GMS2X_amd64.dll

Might want to hear what @niermann thinks about this.

**Comment by thomasaarholt** [comment]
Ah, I didn't spot that part in the readme. I'll attempt again.

Edit: Yeah, works! Still doesn't read the dimensions in in the right order, but still pretty good!

**Comment by ericpre** [comment]
If I remember correctly, you just need to transpose (`s.T` or something similar) the data before saving as hdf5/hspy file. DM can not transpose large dataset, it just crashes...

**Comment by niermann** [comment]
As already discussed the binaries are provided in the release zip. Currently the plugin is just intended to provide methods (accessible by DM scripting) to query HDF5 files, and read datasets and attributes. Also a DM script is provided to read my own internal custom datasets. What's missing here is a DM script, which interprets the data/metadata stored in hyperspy files in a way DM will understand. I'm not (yet) using hyperspy, and also try to avoid DM scripting, so I have no intention to write such a script myself in the next future. However, I'm happy to help anyone, who is willing to write such a script.

**Comment by thomasaarholt** [comment]
@niermann @francisco-dlp 
I've had a bit of a play, and I can confirm that the plugin works great with hyperspy. My previous errors were due to the order of the data dimensions and the fact that my data was in float64.

There are two requirements for the file to import correctly:

1. The data type must not be float64, but float32 (`s.change_dtype("float32")`)
2. The data must be transposed before saving (`s.T.save(filename)` works great)

The first point is only necessary if one wants to use the spectrum picker tool, for EELS and EDS, as far as I can tell.

Finally, I have modified a single string in the import script, to default to importing "*.hspy" files, rather than "*.hdf5". In case we want to put it on the website.
[CMD_import_hspy.zip](https://github.com/hyperspy/hyperspy/files/1115128/CMD_import_hspy.zip)

**Comment by thomasaarholt** [comment]
Currently, the hyperspy metadata, such as units, offset, scale and things like convergence angle are not included. I'll try to hunt them down.

**Comment by thomasaarholt** [comment]
@niermann I'm trying to work out how to look for the attribute "units" located under axis-0 using your script, on the following file.
[EELS.zip](https://github.com/hyperspy/hyperspy/files/1124263/EELS.zip)

I'm using HDFView to understand the layout of the file. Here's what it looks like in this case:

![image](https://user-images.githubusercontent.com/2721423/27859884-8b466f04-617a-11e7-8e9c-de669ff182b0.png)

I suspect it's got something to do with the following lines, although simply replacing "unit" for "units" did nothing. Do you have any advice? I'm afraid I'm not entirely sure how to debug scripts in DM.
```
    // Look for intensity scale
    Number data_scale = 1.0
    if (TagGroupGetTagAsNumber(attr, "scale", data_scale))
        ImageSetIntensityScale(array, data_scale)
    Number data_offset = 0.0
    if (TagGroupGetTagAsNumber(attr, "offset", data_offset))
        ImageSetIntensityOrigin(array, data_offset / data_scale)
    String data_unit
    if (TagGroupGetTagAsString(attr, "unit", data_unit))
        ImageSetIntensityUnitString(array, data_unit)
```

**Comment by niermann** [comment]
I'm currently on a business trip and will look into this if I find some spare time. Debugging in DM is a pain, you can use the Result(...) function, which roughly behaves like a print function. That's it.

**Comment by thomasaarholt** [comment]
*kill me*

**Comment by niermann** [comment]
I'm not really sure were the problem is. I would need to see more of the script you are using. Make sure you read the attributes of the right datasets/group. In the example file "axis-0" is a group. Then use h5_read_attr to read all its attributes into a DM taggroup, and then its method to read a particular attribute/tag. 

The following snippet works for me...

// Get path of file
String filename
if (!OpenDialog(Null, "Select HDF5 file", "*.hdf5", filename))
  exit(0)

// Read attributes of dataset "Experiments/__unnamed__/axis-0"
TagGroup attr
attr = h5_read_attr(filename, "Experiments/__unnamed__/axis-0")

//Use this to view contents of taggroup
//TagGroupOpenBrowserWindow(attr, 0)

// Read attribute "units"
String units
if (TagGroupGetTagAsString(attr, "units", units))
	Result("Units: " + units + "\n")
else
	Result("Error reading units tag\n")

**Comment by k8macarthur** [comment]
Just seen this thread! I'm trying to get data from the FEI velox software into DM (technically HDF5 format but with the extension .emd). Currently I have tested the script on my GMS 3 and it does run without errors. Unfortunately, it slices up my data in wrong directions and I think the actual data cube I'm interested in is stored in the shape of 1xmany when I'm hoping it has 4 dimensions.  I think I'm having the same ordering problem. Is there anyway to transpose in DM scripting easily/  import a different way around?

**Comment by ericpre** [comment]
@k8macarthur, alternatively, you can check #1662 out and use a script to convert the emd file to appropriate hdf5 file.
If you are talking about the stream of EDS data, this is not an usual ordering issue of the dimensions, you can look at #1662, where reading EDS data is documented.

**Comment by thomasaarholt** [comment]
Yeah, I'd do what @ericpre is suggesting and save it in hyperspy as a .hspy, then import it using the DM script (do remember to transpose the data before saving it, in hyperspy).

**Comment by k8macarthur** [comment]
Cool, thanks guys. I'll have a look next week :)

Best wishes,

Katherine MacArthur

**Comment by niermann** [comment]
I have just noticed that DM (only tested for GMS 2.3) has a function, which can transpose data: slice2/slice3.

@thomasaarholt, this might also be useful for a hspy import script, which does not rely on transposing the data before saving it.

The syntax is a bit cumbersome, see the DM documentation for that (the doc in my installation is located in C:\ProgramData\Gatan\Documentation\Scripting_DigitalMicrograph.pdf). If you want to transpose an image "img" with dimensions dim_x=1024 dim_y=20 dim_z=10, corresponding to a numpy shape of (10, 20, 1024), into dim_x=20 dim_y=10 dim_z=1024,  it can be done with:
`ShowImage(slice3(img, 0, 0, 0, 1, 20, 1, 2, 10, 1, 0, 1024, 1))`

Note, that this creates a different "view" (speaking in numpy terms) on the array and does not copy it (so still might work on larger arrays, which otherwise tend to crash DM). Sadly this works only for up to 3 dimensions (there is no slice4, at least in GMS 2.3).

**Comment by tjof2** [comment]
Closed in #2542
